cmd/tealdbg/local.go

```diff
@@ -573,44 +539,19 @@ func (r *LocalRunner) Run() (bool, error) {
 		return false, fmt.Errorf("no program to debug")
 	}
 
-	pooledApplicationBudget := uint64(0)
-	credit, _ := transactions.FeeCredit(r.txnGroup, r.proto.MinTxnFee)
-	// ignore error since fees are not important for debugging in most cases
+	txngroup := transactions.WrapSignedTxnsWithAD(r.txnGroup)
 
-	evalParams := make([]logic.EvalParams, len(r.runs))
-	for i, run := range r.runs {
-		if run.mode == modeStateful {
-			if r.proto.EnableAppCostPooling {
-				pooledApplicationBudget += uint64(r.proto.MaxAppProgramCost)
-			} else {
-				pooledApplicationBudget = uint64(r.proto.MaxAppProgramCost)
-			}
-		}
-		ep := logic.EvalParams{
-			Proto:                   &r.proto,
-			Txn:                     &r.txnGroup[run.groupIndex],
-			TxnGroup:                r.txnGroup,
-			GroupIndex:              run.groupIndex,
-			PastSideEffects:         run.pastSideEffects,
-			Specials:                &transactions.SpecialAddresses{},
-			FeeCredit:               &credit,
-			PooledApplicationBudget: &pooledApplicationBudget,
-		}
-
-		// Workaround for Go's nil/empty interfaces nil check after nil assignment, i.e.
-		// r.debugger = nil
-		// ep.Debugger = r.debugger
-		// if ep.Debugger != nil // FALSE
-		if r.debugger != nil {
-			r.debugger.SaveProgram(run.name, run.program, run.source, run.offsetToLine, run.states)
-			ep.Debugger = r.debugger
-		}
-
-		evalParams[i] = ep
-	}
+	ep := logic.NewEvalParams(txngroup, &r.proto, &transactions.SpecialAddresses{})
 
 	run := r.runs[0]
-	ep := evalParams[0]
+	// Workaround for Go's nil/empty interfaces nil check after nil assignment, i.e.
+	// r.debugger = nil
+	// ep.Debugger = r.debugger
+	// if ep.Debugger != nil // FALSE
+	if r.debugger != nil {
+		r.debugger.SaveProgram(run.name, run.program, run.source, run.offsetToLine, run.states)
+		ep.Debugger = r.debugger
+	}
```

Comment:
this is correct, but it looks like it needs to be applied every time you assign a debugger.
In that case, how do you feel about adding a method `EvalParams.AttachDebugger` that would accomplish that ?
( i.e. testing nil interface as well as nil debugger interface pointer )

Reply:
Although this looks like "my" code in the diff, it's actually just moving some existing code to deal with the way EvalParams are shared across executions in a group.  So I'm going to assume Pavel is right here, and no other problem areas exist.